File: src/UAMotorsCADAlert/Services/UserService.cs
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text;

namespace UAMotorsCADAlert.Services;

public class UserProfile
{
    [JsonPropertyName("email")]
    public string Email { get; set; } = string.Empty;
    
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;
}

public class WhitelistEntry
{
    [JsonPropertyName("nombre")]
    public string Nombre { get; set; } = string.Empty;
}

public class DriveDatabase
{
    [JsonPropertyName("config")]
    public Dictionary<string, string> Config { get; set; } = new();

    [JsonPropertyName("users")]
    public Dictionary<string, WhitelistEntry> Users { get; set; } = new();
}

public static class UserService
{
    public static string GetProfileFilePath() => Path.Combine(Config.GetAppDataDir(), "user_profile.uamotors");

    public static UserProfile? LoadLocalProfile()
    {
        string path = GetProfileFilePath();
        if (!File.Exists(path)) return null;

        try
        {
            byte[] encryptedData = File.ReadAllBytes(path);
            byte[] decryptedData = ProtectedData.Unprotect(encryptedData, null, DataProtectionScope.CurrentUser);
            string json = Encoding.UTF8.GetString(decryptedData);
            return JsonSerializer.Deserialize<UserProfile>(json);
        }
        catch
        {
            return null;
        }
    }

    public static UserProfile SaveLocalProfile(string email, string name)
    {
        // Sanitizacion de caracteres especiales
        string cleanName = name.Trim().ToUpper().Replace("*", "").Replace("_", "").Replace("`", "").Replace("~", "");
        var profile = new UserProfile { Email = email.Trim().ToLower(), Name = cleanName };
        
        string jsonStr = JsonSerializer.Serialize(profile);
        byte[] dataToEncrypt = Encoding.UTF8.GetBytes(jsonStr);
        byte[] encryptedData = ProtectedData.Protect(dataToEncrypt, null, DataProtectionScope.CurrentUser);
        
        File.WriteAllBytes(GetProfileFilePath(), encryptedData);
        return profile;
    }

    private static byte[] GetEncryptionKey()
    {
        string secret = "UAMOTORS_2026_CAD_ELECTRONICS";
        string saltStr = "UAMOTORS_CAD_ALERT_SALT";
        byte[] salt = Encoding.UTF8.GetBytes(saltStr);
        
        using var pbkdf2 = new Rfc2898DeriveBytes(secret, salt, 100000, HashAlgorithmName.SHA256);
        return pbkdf2.GetBytes(32);
    }

    private static byte[] Base64UrlDecode(string input)
    {
        string base64 = input.Replace('-', '+').Replace('_', '/');
        switch (base64.Length % 4)
        {
            case 2: base64 += "=="; break;
            case 3: base64 += "="; break;
        }
        return Convert.FromBase64String(base64);
    }

    public static DriveDatabase? LoadDriveDatabase(string rutaUAMOTORS)
    {
        string dbPath = Path.Combine(rutaUAMOTORS, Config.RelDriveDbPath);
        if (!File.Exists(dbPath)) return null;

        try
        {
            byte[] encryptedBytes = File.ReadAllBytes(dbPath);
            string fernetToken = Encoding.UTF8.GetString(encryptedBytes);
            byte[] tokenBytes = Base64UrlDecode(fernetToken);
            
            // Extraccion de componentes de encriptacion
            byte[] iv = new byte[16];
            Array.Copy(tokenBytes, 9, iv, 0, 16);
            
            int ciphertextLength = tokenBytes.Length - 9 - 16 - 32;
            byte[] ciphertext = new byte[ciphertextLength];
            Array.Copy(tokenBytes, 25, ciphertext, 0, ciphertextLength);

            byte[] key = GetEncryptionKey(); 
            // Division de llave para descifrado AES
            byte[] encKey = new byte[16];
            Array.Copy(key, 16, encKey, 0, 16); 

            using var aes = Aes.Create();
            aes.Key = encKey;
            aes.IV = iv;
            aes.Mode = CipherMode.CBC;
            aes.Padding = PaddingMode.PKCS7;

            using var decryptor = aes.CreateDecryptor();
            byte[] decryptedBytes = decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
            string json = Encoding.UTF8.GetString(decryptedBytes);
            
            return JsonSerializer.Deserialize<DriveDatabase>(json);
        }
        catch
        {
            return null;
        }
    }

    public static (bool Success, string? Name, string ErrorMsg) VerifyUserEmail(string email, string rutaUAMOTORS)
    {
        string emailClean = email.Trim().ToLower();
        if (string.IsNullOrEmpty(emailClean) || !emailClean.Contains("@"))
            return (false, null, "Por favor ingresa un correo electrónico válido.");

        var db = LoadDriveDatabase(rutaUAMOTORS);
        if (db == null)
            return (false, null, $"No se encontró la Base de Datos en Drive. Asegúrate de tener Drive abierto.");

        if (db.Users.TryGetValue(emailClean, out var userEntry))
        {
            string name = !string.IsNullOrEmpty(userEntry.Nombre) ? userEntry.Nombre : "Usuario Autorizado";
            return (true, name, string.Empty);
        }

        return (false, null, "Correo no encontrado. Contacta al administrador si crees que es un error.");
    }
}

File: src/services/user_service.py
"""
User authorization whitelist and local profile service.
"""

import os
import json
import base64
import src.config as config


def get_profile_file_path():
    """Returns absolute path to local profile (.uamotors)."""
    return os.path.join(config.get_appdata_dir(), "user_profile.uamotors")


def load_local_profile():
    """Reads and decodes user_profile.uamotors if it exists, otherwise returns None."""
    profile_path = get_profile_file_path()

    if os.path.exists(profile_path):
        try:
            with open(profile_path, "rb") as f:
                encoded_bytes = f.read()
            json_str = base64.b64decode(encoded_bytes).decode("utf-8")
            return json.loads(json_str)
        except Exception:
            return None

    return None


def save_local_profile(email, name):
    """Saves verified user profile encoded in base64 inside user_profile.uamotors."""
    profile_path = get_profile_file_path()

    data = {
        "email": email.strip().lower(),
        "name": name.strip(),
    }

    json_str = json.dumps(data, ensure_ascii=False, indent=2)
    encoded_bytes = base64.b64encode(json_str.encode("utf-8"))

    with open(profile_path, "wb") as f:
        f.write(encoded_bytes)

    return data


def load_drive_whitelist(ruta_uamotors):
    """Reads and decodes authorized_users.uamotors from Google Drive."""
    db_path = os.path.join(ruta_uamotors, config.REL_DRIVE_DB_PATH)

    if not os.path.exists(db_path):
        return None

    try:
        with open(db_path, "rb") as f:
            encoded_bytes = f.read()
        json_str = base64.b64decode(encoded_bytes).decode("utf-8")
        return json.loads(json_str)
    except Exception:
        return None


def verify_user_email(email, ruta_uamotors):
    """
    Verifies if an email is valid and registered in the Drive whitelist.
    
    Returns:
        tuple: (bool success, str name_or_none, str error_msg)
    """
    email_clean = email.strip().lower()
    if not email_clean or "@" not in email_clean:
        return False, None, "Por favor ingresa un correo electrónico válido."

    whitelist = load_drive_whitelist(ruta_uamotors)
    if whitelist is None:
        return (
            False,
            None,
            f"No se encontró la BD de usuarios en Drive ({config.REL_DRIVE_DB_PATH}).",
        )

    if email_clean in whitelist:
        user_info = whitelist[email_clean]
        name = user_info.get("nombre", "Usuario Autorizado")
        return True, name, ""
    else:
        return (
            False,
            None,
            "El correo no está registrado en la lista de usuarios autorizados de UAMOTORS.",
        )

